cdap-security/src/main/java/co/cask/cdap/security/store/FileSecureStore.java

```diff
@@ -0,0 +1,463 @@
+      // If another flush does not happen and the system is restarted, the only time that file is read,
+      // then we will have an extra key.
+      flush();
+    } catch (IOException ioe) {
+      if (key != null) {
+        try {
+          keyStore.setKeyEntry(name, key, password, null);
+        } catch (KeyStoreException e) {
+          throw new IOException("Failed to put back the key after a flush failure.", e);
+        }
+      }
+      throw ioe;
+    } catch (UnrecoverableKeyException | NoSuchAlgorithmException | KeyStoreException e) {
+      e.printStackTrace();
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  /**
+   * List of all the entries in the secure store.
+   * @return A list of {@link SecureStoreMetadata} objects representing the data stored in the store.
+   */
+  @Override
+  public List<SecureStoreMetadata> list() throws IOException {
+    readLock.lock();
+    try {
+      Enumeration<String> aliases = keyStore.aliases();
+      String name;
+      List<SecureStoreMetadata> list = new ArrayList<>();
+      while (aliases.hasMoreElements()) {
+        name = aliases.nextElement();
+        list.add(getSecureStoreMetadata(name));
+      }
+      return list;
+    } catch (KeyStoreException e) {
+      throw new IOException("Failed to get the list of elements from the secure store.", e);
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  /**
+   * @param name Name of the data element.
+   * @return An object representing the securely stored data associated with the name.
+   */
+  @Override
+  public SecureStoreData get(String name) throws IOException {
+    readLock.lock();
+    try {
+      if (!keyStore.containsAlias(name)) {
+        throw new IOException(name + " not found in the secure store.");
+      }
+      Key key = keyStore.getKey(name, password);
+      return ((KeyStoreEntry) key).getData();
+    } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException e) {
+      throw new IOException("Unable to retrieve the key " + name, e);
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  /**
+   * Returns the metadata for the element identified by the given name.
+   * @param name Name of the element
+   * @return An object representing the metadata associated with the element
+   */
+  private SecureStoreMetadata getSecureStoreMetadata(String name) throws IOException {
+    readLock.lock();
+    try {
+      if (!keyStore.containsAlias(name)) {
+        throw new IOException("Metadata for " + name + " not found in the secure store.");
+      }
+      Key key = keyStore.getKey(name, password);
+      return ((KeyStoreEntry) key).metadata;
+    } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException e) {
+      throw new IOException("Unable to retrieve the metadata for " + name, e);
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  private static Path constructNewPath(Path path) {
+    return path.resolveSibling(path + "_NEW");
```

Comment:
path + "_NEW" should also work, if the argument is an absolute path then it is simply returned.
I have changed it to path.getFileName() + "_NEW" to make it easier to understand.